File: Day_04/solution.pl
#!/opt/perl/bin/perl

use 5.028;

use strict;
use warnings;
no  warnings 'syntax';

use experimental 'signatures';
use experimental 'lexical_subs';

package BingoCard {
    use Hash::Util::FieldHash qw [fieldhash];
    fieldhash my %card;
    fieldhash my %size;
    fieldhash my %finished;

    sub new ($class) {bless do {\my $var} => $class}
    sub init ($self, $input) {
        $card {$self} = [map {[/[0-9]+/g]} split /\n/ => $input];
        $size {$self} = @{$card {$self}};
        $self
    }
    sub bingo ($self) {
        my $card = $card {$self};
        my $size = $size {$self};
        foreach my $r (0 .. $size - 1) {
            return 1 unless grep {$_ >= 0} @{$$card [$r]};
        }
        foreach my $c (0 .. $size - 1) {
            return 1 unless grep {$_ >= 0}
                             map {$$card [$_] [$c]} 0 .. $size - 1;
        }
        return 0;
    }
    sub play ($self, $number) {
        foreach my $row (@{$card {$self}}) {
            foreach my $i (keys @$row) {
                $$row [$i] = -1 if $$row [$i] == $number;
            }
        }
        $self
    }
    sub left ($self) {
        my $sum = 0;
        foreach my $row (@{$card {$self}}) {
            foreach my $num (@$row) {
                $sum += $num if $num >= 0;
            }
        }
        $sum
    }
    sub finished ($self) {
        $finished {$self}
    }
    sub set_finished ($self) {
        $finished {$self} = 1
    }
}

@ARGV = "input" unless @ARGV;
$/ = "";  # Paragraph mode

my @numbers = <> =~ /[0-9]+/g;

my @cards = map {BingoCard:: -> new -> init ($_)} <>;

my $first_win = 0;
my $last_win  = 0;

foreach my $number (@numbers) {
    foreach my $i (keys @cards) {
        my $card = $cards [$i];
        next if $card -> finished;
        if ($card -> play ($number) -> bingo) {
            $first_win ||= $number * $card -> left;
            $last_win    = $number * $card -> left;
            $card -> set_finished;
        }
    }
}

say "Solution 1: $first_win";
say "Solution 2: $last_win";


__END__

File: Day_04/solution.py
class BingoCard:

    def __init__ (self, card):
        num_dict = {}
        row_list = []
        col_list = []
        sum      =  0
        lines = card . rstrip () . split ("\n")
        rc = -1
        for line in lines:
            rc = rc + 1
            row_list . append (0)
            nums = line . strip () . split ()
            cc = -1
            for num in nums:
                cc = cc + 1
                if rc == 0:
                    col_list . append (0)
                num = int (num)
                num_dict [num] = rc, cc
                row_list [rc]  = row_list [rc] + 1
                col_list [cc]  = col_list [cc] + 1
                sum            = sum + num

        self . numbers = num_dict
        self . rows    = row_list
        self . cols    = col_list
        self . bingo   = 0
        self . total   = sum  

    def play (self, number):
        if self . bingo == 0:
            if number in self . numbers:
                [rc, cc] = self . numbers [number]
                self . rows [rc] = self . rows [rc] - 1
                self . cols [cc] = self . cols [cc] - 1
                if self . rows [rc] == 0 or self . cols [cc] == 0:
                    self . bingo = 1
                self . total = self . total - number


input = open ("input", mode = 'r') . read () . split ("\n\n")
balls = input [0]
cards = map (lambda sheet: BingoCard (sheet), input [1:])

first_score = -1
last_score  = -1
for ball in map (lambda b: int (b), balls . split (",")):
    for card in cards:
        card . play (ball)
        if card . bingo:
            if first_score < 0:
                first_score = ball * card . total
            last_score = ball * card . total
    cards = filter (lambda card: card . bingo == 0, cards)

print ("Solution 1: %d" %(first_score))
print ("Solution 2: %d" %(last_score))
